protocol-test-util/src/main/java/io/camunda/zeebe/test/broker/protocol/record/ProtocolFactory.java

```diff
@@ -0,0 +1,274 @@
+  }
+
+  private EasyRandomParameters getDefaultParameters() {
+    // as we will ensure value/intent/valueType having matching types, omit them from the
+    // randomization process - we will do that individually afterwards
+    final Predicate<Field> excludedRecordFields =
+        FieldPredicates.inClass(ImmutableRecord.class)
+            .and(
+                FieldPredicates.named("value")
+                    .or(FieldPredicates.named("intent"))
+                    .or(FieldPredicates.named("valueType")));
+
+    return new EasyRandomParameters()
+        .randomizerRegistry(randomizerRegistry)
+        // we have to bypass the setters since our types are almost exclusively immutable
+        .bypassSetters(true)
+        // allow empty collections, and only generate up to 5 items
+        .collectionSizeRange(0, 5)
+        // as we have nested types in our protocol, let's give a generous depth here, but let's
+        // still limit it to avoid errors/issues with nested collections
+        .randomizationDepth(8)
+        .excludeField(excludedRecordFields);
+  }
+
+  private Record<RecordValue> generateImmutableRecord(
+      final ValueType valueType, final UnaryOperator<Builder<RecordValue>> modifier) {
+    Objects.requireNonNull(valueType, "must specify a value type");
+    Objects.requireNonNull(modifier, "must specify a builder modifier");
+
+    final var typeInfo = ValueTypeMapping.get(valueType);
+    final var intent = random.nextObject(typeInfo.getIntentClass());
+    final var value = generateObject(typeInfo.getValueClass());
+    final var seedRecord = random.nextObject(Record.class);
+
+    //noinspection unchecked
+    final Builder<RecordValue> builder =
+        ImmutableRecord.builder()
+            .from(seedRecord)
+            .withValueType(valueType)
+            .withValue(value)
+            .withIntent(intent);
+
+    return Objects.requireNonNull(modifier.apply(builder), "must return a non null builder")
+        .build();
+  }
+
+  /**
+   * A few of the protocol types refer to {@code Map<String, Object>}. easy-random doesn't really
+   * know how to randomize the object type, and instead just creates raw {@link Object} instances.
+   * This not only doesn't really simulate the expected data set, but it's not serializable, which
+   * would prevent this class from being used in any kind of serialization context (e.g. exporters)
+   *
+   * <p>To circumvent this, we randomize the type to some easy to serialize and randomize type, and
+   * return that.
+   *
+   * <p>NOTE: floats and doubles were omitted on purpose as Jackson has trouble deserializing them,
+   * and instead will always deserialize them as {@link java.math.BigDecimal}. This is a regression,
+   * so we could think about adding them back once that's fixed, but it didn't seem super important
+   * at the moment.
+   */
+  private final class RawObjectRandomizer implements Randomizer<Object> {
```

Comment:
❓ Why is this an inner class, while EnumRandomizer is a standalone class? If there's no clear reason, please align for consistency.

Reply:
`RawObjectRandomizer` is not static, and reuses the `random` instance of the enclosing class. This could admittedly be simplified to a standalone randomizer, as we return primitives (so no need for easy-random here, we can use a normal random). In fact, I think I'll do that :+1:

Reply:
I tried it, and I will stick with the current solution. The main reason is that reusing the random instance here means it will be configured the same, and it will reuse the same parameters. We can more easily extend it in the future.